utils/big_decimal.cc

```diff
@@ -143,6 +143,99 @@ std::strong_ordering big_decimal::operator<=>(const big_decimal& other) const
     return x.compare(y) <=> 0;
 }
 
+std::strong_ordering big_decimal::operator<=>(const big_decimal& other) const
+{
+    if (_scale == other._scale) {
+        return _unscaled_value.compare(other._unscaled_value) <=> 0;
+    }
+
+    const int sign = boost::multiprecision::sign(_unscaled_value) < 0 ? -1 : 1;
+    const int sign_other = boost::multiprecision::sign(other._unscaled_value) < 0 ? -1 : 1;
+    if (sign != sign_other) {
+        return sign <=> sign_other;
+    }
+
+    const int is_zero = !_unscaled_value;
+    const int is_zero_other = !other._unscaled_value;
+    if (is_zero != is_zero_other) {
+        // We know the numbers have the same sign and one of them is zero, so they are both positive numbers.
+        return !is_zero <=> !is_zero_other;
+    } else if (is_zero) {
+        // Both are zero
+        return std::strong_ordering::equal;
+    }
+
+    // At this point we know that both numbers have the same sign, so if one is negative, the other is too.
+    const bool comparing_negatives = (sign < 0);
+
+    // At this point we know that the numbers:
+    // * have different scale
+    // * have the same sign
+    // * neither is zero
+    //
+    // The numbers have the form:
+    //
+    //      auto number = _unscaled_value * std::pow(10, -_scale);
+    //      auto number_other = other._unscaled_value * std::pow(10, -other._scale);
+    //
+    // To compare them, we make the unscaled values the same (or close), so we can directly compare the scales.
+    // To do that we want to compute diff_scale_2:
+    //
+    //      auto unscaled_diff = _unscaled_value / other._unscaled_value;
+    //      auto diff_scale_2 = log2(unscaled_diff);
+    //
+    // To avoid using division and then calculating a log2(), we use the number
+    // of bits in both numbers do infer log2(unscaled_diff) directly.
+    const int64_t bit_count = boost::multiprecision::msb(comparing_negatives ? -_unscaled_value : _unscaled_value);
+    const int64_t bit_count_other = boost::multiprecision::msb(comparing_negatives ? -other._unscaled_value : other._unscaled_value);
+    const int64_t diff_scale_2 = bit_count - bit_count_other;
+
+    // Now we can rewrite the original numbers as follows:
+    //
+    //      auto number = _unscaled_value * std::pow(10, -_scale);
+    //      auto number_other_approx = _unscaled_value * std::pow(2, diff_scale_2) * std::pow(10, -other._scale);
+    //
+    // Notice that number_other_approx != number_other, but it is close, the following holds:
```

Comment:
It would be easier to notice it if we had the formula for number_other to compare to.

Reply:
In v4 I added here the definition of `number_other` too, to make this easier to follow.